### PR TITLE
Fix fd reuse errors with plasma fallback allocation

### LINKED ISSUE #16460: Possible memory corruption on fd reuse if plasma_unlimited=1
The plasma client does not handle re-use of file descriptors properly. Re-use of a file descriptor id can cause memory corruption as the client tries to map memory from the wrong files.

FD re-use can occur if we ever un-map a file, which occurs when we deallocate filesystem-backed pages with RAY_plasma_unlimited=1.

This issue was discovered in https://github.com/ray-project/ray/pull/16451 (thanks @scv119 !)

```
import random
import ray
import numpy as np

@ray.remote
class Actor:
    def produce(self, i):
        s = int(random.random() * 200)
        z = np.ones(s * 1024 * 1024)
        z[0] = i
        return z

    def consume(self, x):
        print(x)

ray.init(object_store_memory=100e6)
i = 0
a = Actor.remote()
b = Actor.remote()
while True:
    x_id = a.produce.remote(i)
    i += 1
    ray.get(b.consume.remote(x_id))
```

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

With RAY_plasma_unlimited mode, allocations that can't fit in /dev/shm get placed in separate filesystem files. These files are removed by dlmalloc once the object is deallocated. However, this deallocation causes fd ids to get re-used, which wasn't handled properly in plasma.

It is desirable to deallocate these filesystem files in order to prevent allocations from going to the filesystem instead of /dev/shm once memory becomes available.

This fixes https://github.com/ray-project/ray/issues/16460

### CODE REVIEW

File: python/ray/tests/test_plasma_unlimited.py [hunk 1]
```diff
@@ -171,6 +172,29 @@ def consume(refs):
         ray.shutdown()
 
 
+@pytest.mark.skipif(
+    platform.system() == "Windows", reason="Need to fix up for Windows.")
+def test_fd_reuse_no_memory_corruption(shutdown_only):
```
Comment (ericl) [inline python/ray/tests/test_plasma_unlimited.py:177]: This test fails on master.

File: src/ray/object_manager/plasma/compat.h [hunk 2]
```diff
@@ -40,14 +41,18 @@ mach_port_t pthread_mach_thread_np(pthread_t);
...
-#define MEMFD_TYPE int
+#define MEMFD_TYPE_NON_UNIQUE int
 #define INVALID_FD -1
 #define FD2INT(x) (x)
 #define INT2FD(x) (x)
 #endif  // #ifndef _WIN32
+
+// Pair of (fd, unique_id). We need to separately track a unique id here
+// since fd values can get re-used by the operating system.
+#define MEMFD_TYPE std::pair<MEMFD_TYPE_NON_UNIQUE, int64_t>
```
Comment (ericl) [inline src/ray/object_manager/plasma/compat.h:58]: Pairs are conveniently hashable, but open to other suggestions here.

File: src/ray/object_manager/plasma/plasma.fbs [hunk 3]
```diff
@@ -177,6 +179,8 @@ table PlasmaCreateReply {
   // The file descriptor in the store that corresponds to the file descriptor
   // being sent to the client right after this message.
   store_fd: int;
+  // The unique id of the store file descriptor in case of fd reuse.
+  unique_fd_id: int;
```
Comment (scv119) [inline src/ray/object_manager/plasma/plasma.fbs:183]: should this be long?

File: src/ray/object_manager/plasma/store.cc [hunk 4]
```diff
@@ -279,7 +281,8 @@ uint8_t *PlasmaStore::AllocateMemory(size_t size, MEMFD_TYPE *fd, int64_t *map_s
 
   if (pointer != nullptr) {
     GetMallocMapinfo(pointer, fd, map_size, offset);
-    RAY_CHECK(*fd != INVALID_FD);
+    RAY_CHECK(fd->first != INVALID_FD);
+    RAY_CHECK(fd->second != 0);
```
Comment (scv119) [inline src/ray/object_manager/plasma/store.cc:285]: nit: maybe create a constant for INVALID_UNIQUE_ID = 0

**Review by scv119 — COMMENTED** [review]
looks good to me but I might not have the full context. @suquark please take a look :)

File: src/ray/object_manager/plasma/malloc.cc [hunk 5]
```diff
@@ -36,13 +36,15 @@ void GetMallocMapinfo(void *addr, MEMFD_TYPE *fd, int64_t *map_size, ptrdiff_t *
...
+      fd->first = entry.second.fd.first;
+      fd->second = entry.second.fd.second;
       *map_size = entry.second.size;
       *offset = pointer_distance(entry.first, addr);
       return;
     }
   }
-  *fd = INVALID_FD;
+  fd->first = INVALID_FD;
+  fd->second = 0;
```
Comment (scv119) [inline src/ray/object_manager/plasma/malloc.cc:47]: INVALID_UNIQUE_FD_ID

File: src/ray/object_manager/plasma/store.cc [hunk 6]
```diff
@@ -338,7 +341,7 @@ PlasmaError PlasmaStore::CreateObject(const ObjectID &object_id,
     return PlasmaError::ObjectExists;
   }
 
-  MEMFD_TYPE fd = INVALID_FD;
+  MEMFD_TYPE fd{INVALID_FD, 0};
```
Comment (scv119) [inline src/ray/object_manager/plasma/store.cc:344]: INVALID_UNIQUE_FD_ID

**Review by scv119 — APPROVED** [review]

**Pull request merged**